File: 230. Kth Smallest Element in a BST/Solution.go
package main

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
	idx := 0
	smallest := 0
	inOrderTraversal(root, k ,&idx, &smallest)
	return smallest
}

func inOrderTraversal( root * TreeNode ,k int ,idx * int , smallest *int )  {
	if root == nil {
		return
	}
	inOrderTraversal(root.Left,k, idx, smallest)
	*idx ++
	if *idx == k {
		*smallest = root.Val
		return
	}

	inOrderTraversal(root.Right,  k,idx, smallest)
}

File: 230. Kth Smallest Element in a BST/Solution.java



class Solution {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    int idx = 0;
    int smallest = 0;
    
    public int kthSmallest(TreeNode root, int k) {
        inOrderTraversal(root, k);
        return smallest;
    }

    private void inOrderTraversal(TreeNode root , int k) {
        if (root == null) {
            return;
        }
        inOrderTraversal(root.left, k);
        idx++;
        if (idx == k) {
            smallest = root.val;
            return;
        }
        inOrderTraversal(root.right, k);   
    }
}